GenMat.cpp

```diff
@@ -0,0 +1,552 @@
+  {
+      for (int i = 0; i < A.height; i++) 
+      {
+       for (int j = 0; j < A.width; j++) {
+        std::cout << A.elements[i * A.width + j];
+        std::cout << " ";
+      }
+      std::cout << std::endl;
+     }
+     std::cout << std::endl;
+     std::cout << std::endl;
+
+
+     dim3 dimBlockTransA(gcd(aT.width,aT.height),gcd(aT.width,aT.height));
+     dim3 dimGridTransA(aT.width/dimBlockTransA.x, aT.height/dimBlockTransA.y);
+     Transpose<<<dimGridTransA,dimBlockTransA>>>(dA,dAt);
+     aT.load(dAt);
+
+     std::cout << "Transpose" << std::endl;
+     for (int i = 0; i < aT.height; i++) 
+     {
+       for (int j = 0; j < aT.width; j++) {
+        std::cout << aT.elements[i * aT.width + j];
+        std::cout << " ";
+      }
+      std::cout << std::endl;
+     }
+     std::cout << std::endl;
+     std::cout << std::endl;
+  }
+  else
+  {
+     dim3 dimBlockTransA(gcd(A.width,A.height),gcd(A.width,A.height));
+     dim3 dimGridTransA(A.width/dimBlockTransA.x, A.height/dimBlockTransA.y);
+     GpuCopy<<<dimGridTransA, dimBlockTransA>>>(dA,dAt);
+     aT.load(dAt);
+  }
+
+
+  // B transpose
+  if(opB)
+  {
+
+     //CMat bT(B.height,B.width);
+     //GMat dBt(B.height,B.width);
+
+     dim3 dimBlockTransB(gcd(bT.width,bT.height),gcd(bT.width,bT.width));
+     dim3 dimGridTransB(bT.width/dimBlockTransB.x, bT.height/dimBlockTransB.y);
+     Transpose<<<dimGridTransB,dimBlockTransB>>>(dB,dBt);
+     bT.load(dBt);
+  }
+  else
+  {
+     //CMat bT(B.width,B.height);
+     //GMat dBt(B.width,B.height);
+     dim3 dimBlockTransB(33/2,33/2);
+     //dim3 dimBlockTransB(gcd(B.width,B.height),gcd(B.width,B.height));
+     std::cout << gcd(B.width, B.height);
+     dim3 dimGridTransB(B.width/dimBlockTransB.x+1, B.height/dimBlockTransB.y+1);
+     GpuCopy<<<dimGridTransB, dimBlockTransB>>>(dB,dBt);
+     
+     bT.load(dBt);
+
+     std::cout << "B" << std::endl;
+     for (int i = 0; i < bT.height; i++) 
+     {
+       for (int j = 0; j < bT.width; j++) {
+        std::cout << bT.elements[i * bT.width + j];
+        std::cout << " ";
+      }
+      std::cout << std::endl;
+     }
+     std::cout << std::endl;
+     std::cout << std::endl;
+  }
+
+  // Multiply A by alpha
+  int blockSize = gcd(aT.width, aT.height);
+  dim3 dimBlock0(blockSize, blockSize);
+  
+  //dim3 dimBlock0(gcd(aT.width,aT.height),gcd(aT.width,aT.height));
+  dim3 dimGrid0(aT.width/dimBlock0.x, aT.height/dimBlock0.y);
+  
+  MatByFloat<<<dimGrid0,dimBlock0>>>(dAt,alpha,dAt);
+  aT.load(dAt);
+
+  std::cout << dimBlock0.x<< " " << dimGrid0.x << " " << dimGrid0.y << std::endl;
+  std::cout << "alpha* A" << std::endl;
+  for (int i = 0; i < aT.height; i++) 
+    {
+    for (int j = 0; j < aT.width; j++) {
+       std::cout << aT.elements[i * aT.width + j];
+       std::cout << " ";
+    }
+    std::cout << std::endl;
+  }
+   std::cout << std::endl;
+   std::cout << std::endl;
+
+
+  // Multiply A and B
+
+  
+  blockSize = gcd(AbyB.width, AbyB.height);
```

Comment:
Again, I wouldn't use the gcd, there are instances where the gcd will be 1, and then the shared memory multiplication reduces to the naive kernel.